index-cn.bs

```diff
@@ -0,0 +1,1116 @@
+
+例如，在谷歌的V8 6.4上(包含在谷歌Chrome 64中)，message是
+
+<pre class=highlight highlight="">
+TypeError: String.prototype.substring called on null or undefined
+</pre>
+
+在Mozilla Firefox 57.0 上给出了更少的有用信息
+
+<pre class=highlight highlight="">
+TypeError: can't convert undefined to object
+</pre>
+
+而与此同时，ChakraCore 1.7.5.0版本(微软Edge中的JavaScript引擎)采用了V8的路线却抛出了
+
+<pre class=highlight highlight="">
+TypeError: String.prototype.substring: 'this' is null or undefined
+</pre>
+
+</div>
+
+## Example: Can <code class=idl>Boolean()</code> and <code class=idl>String()</code> 会抛出异常么? ## {#example-can-boolean-and-string-ever-throw-exceptions}
+
+当在编写任务中重要且关键的代码时，我们通常会将异常处理会放在编码的首要位置。所以，*“我现在用的这个内置函数是不是会抛出异常呢?”* 这样的问题，我们经常会考虑。
+
+在这个例子中，我们会通过 <code class=idl>Boolean()</code> and <code class=idl>String()</code>. 这两个语言内置函数来尝试回答这样的问题。我们将只关注对这两个函数的直接调用，而不是 `new Boolean()` and `new String()` 种包装函数的情况（这种包装函数是JavaScript中最不受欢迎的特性之一，几乎也是所有JS指南都不鼓励的做法） [[YDKJS]].
+
+在导航到规范中的 {{ECMAScript/Boolean()}} 部分后，我们可以看到相当短的算法:
+
+<div algorithm="to execute Boolean()">
+
+When <code class=idl>Boolean</code> is called with argument |value|, the following steps are taken:
+
+1.  Let |b| be ! [$ECMAScript/ToBoolean$](|value|).
+1.  If NewTarget is **undefined**, return |b|.
+1.  Let |O| be ? [$ECMAScript/OrdinaryCreateFromConstructor$](NewTarget, **"%BooleanPrototype%"**, « \[[BooleanData]] »).
+1.  Set |O|.\[[BooleanData]] to |b|.
+1.  Return |O|.
+
+</div>
+
+然而看起来好像并没有那么简单，涉及到了一些关于 [$OrdinaryCreateFromConstructor$] 的复杂技巧。另外，更重要的是在第3步中还有一个[=?=]的简写，这表明这个函数在某些情况下还有可能会抛出错误。让我们来仔细看看吧。 
+
+Step 1 将 |value| (the function argument) 强制转换为Boolean值。有趣的是，这一步没有 [=?=] or [=!=] s简写，但没有一个Completion Record的缩写通常意味着要加一个 [=!=]. 所以*步骤 1*不能抛出异常。
+
+Step 2 检查 [$NewTarget$] is **undefined**. <dfn abstract-op>NewTarget</dfn> 与 {{MDN/new.target}} 在规范上是等价的，目的是让规范能够区分是 `new Boolean()` call (where it is `Boolean`) 还是 `Boolean()` 的调用 (where it is **undefined**).因为我们现在只需要关注对 `Boolean()` 的直接调用，所以我们知道NewTarget的值总是**undefined**，且算法不需要任何额外的处理直接返回*b*。  
+
+因为不需要考虑 `new` 调用的情况，所以当调用 `Boolean()` 时只会访问 `Boolean()` 算法中的前两个步骤，而这两个步骤都不会抛出异常，所以我们得出结论，无论输入是什么，**Boolean()都不会抛出异常**。  
+
+----
+
+OK，让我们把注意力再转向 {{ECMAScript/String()}}:
+
+<div algorithm="to execute String()">
+
+当用 |value| 调用 <code class=idl>String</code> 时:
+
+1.  如果调用这个函数时没有传递参数, 那就将 |s| 设为 `""`.
+1.  Else,
+  1.  If NewTarget is **undefined** and [$ECMAScript/Type$](|value|) is Symbol, return [$ECMAScript/SymbolDescriptiveString$](|value|).
+  1.  Let |s| be ? [$ECMAScript/ToString$](|value|).
+1.  If NewTarget is **undefined**, return |s|.
+1.  Return ? [$ECMAScript/StringCreate$](|s|, ? [$ECMAScript/GetPrototypeFromConstructor$](NewTarget, `"%StringPrototype%"`)).
+
+</div>
+
+根据分析 {{ECMAScript/Boolean()}} 函数的相关经验，我们可以知道 [$NewTarget$] 也始终应该为**undefined**，因此最后一步可以跳过。同时我们还知道 [$Type$] and [$SymbolDescriptiveString$] 也是安全的因为它们都不会处理abrupt completions。然而，在调用[$ToString$] [=abstract operation=]前有一个 [=?=] . 让我们来仔细看看吧。
```

Comment:
Is 处理 the right word here? It's more that they would never **produce** abrupt completions, not that they would not deal with them.

Reply:
fixed it . I thought this 'handled' means like 'handled the situation'

Reply:
Ah oops, I realize that I misunderstood what I meant initially. I did actually mean handle – but more like the _caller_ didn't handle any abrupt completions since they didn't have ? or ! in front of `Type` or `SymbolDescriptiveString`. The original translation made it sound like "`Type` and `SymbolDescriptiveString` are safe, since *they* don't handle abrupt completions."

Could you address this? afterwards it should be ready to merge.

Reply:
OK , now i see the difference. And I think your chinese is probably better then mine 😂.